Fully-automated release process

cachedout opened the pull request:
## Summary

This PR introduces a fully-automated release process.

## Dependencies

This PR cannot be merged until https://github.com/elastic/infra/pull/18234 is merged. Additionally, a new release of the APM Pipeline Library is needed which includes this PR: https://github.com/elastic/apm-pipeline-library/pull/413

## Design

The design of this PR is to be purely additive. 

The current release process can still be executed using the same means as prior to this change. Where previously, however, there were a number of manual steps, this new pipeline can be run which will attempt to automate the bulk of those.

## Use

After the pre-requisite PRs are merged and this PR is also merged, the internal CI server should have a new "post-release" job which contains the logic in this PR.

To run the post-build steps, the job will need to be run. It is important that when launching the job that a window is kept open because the job will present dialogue asking questions as the release process progresses. I would prefer if we perhaps set up a Zoom call during the first few invocations of this job so that I can be on-hand to help debug any errors. While this is tested wherever possible, the nature of this job makes testing certain elements somewhat challenging.

**For the time being, your release process should not change. We will take this as incrementally as you like.**

If there is a desire to split this into more discrete changes, or even to merge this in pieces, I am open to that option. Please provide feedback about this question if it's a concern.

## Maturity

This is being submitted as a draft in order to get feedback from the AQA team and from the apm-agent-java folks.

## Reviewing

### For the AQA team:

 * Please make sure the Jenkins pipeline syntax looks correct. I'm sure there are a few places where things could be improved.

* In a few cases, this drops back to a shell. Please mark any cases where you think this would be better encapsulated in our pipeline library or where the same functionality can be achieved using existing Jenkins run-time variables or by composing pieces of our library. (I'm almost certain there are at least one or two of these.

* Bugs :)

### For the apm-agent-java team:

* I tried my best to leave plain-text documentation to make this easy to follow. In a few cases, it may not be obvious what is happening so please ask questions!

* If you don't want to automate any of these steps, it's very easy to put them behind gates and default them to being disabled. (Or just to remove them entirely.) Please mark any areas of concerns where you might like us to do this.

* I know that we discussed the automated CHANGELOG in another issue. I have included a step here which generates a CHANGELOG and just provides it as an artifact at the end of the build. I am happy to remove/disable it if you aren't interested in having this artifact available to inspect. It does not modify files in the repo or affect the build in any way.

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+                  def newBranchName = sh(script: 'git tag|tail -1|sed s/v//|cut -f1 -d "."|awk \'{print $1".x"}\'', returnStdout: true)
+                  echo "This appears to be a major version. We are creating a new branch for $newBranchName in the apm-agent-java repo on GitHub."
+                  // Now we need to branch
+                  sh(script: "git checkout -b $newBranchName")
```
What do you think to move this logic to a shell script?

SylvainJuge replied:
Also, as an alternative, it might be easier to get project version from maven `pom.xml` rather than parsing it from git.

cachedout replied:
Great idea! I have [implemented this upstream](https://github.com/elastic/apm-pipeline-library/pull/432) and will update this pipeline to use the new functionality.

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+                def baseUrl = "https://www.elastic.co/guide/en/apm/agent/java/current/release-notes-"
+                def dotX = sh(script: 'git tag|tail -1|sed s/v//|cut -f1 -d "."|awk \'{print $1".x"}\'', returnStdout: true)
+                def bareTag = sh(script: 'git tag|tail -1|sed s/v//', returnStdout: true)
+                def finalUrl = sh(script: "echo $baseUrl$dotX\.html#release-notes-$bareTag")
```
Shell script that returns the URL?

cachedout replied:
Good suggestion! I have [migrated this logic to a stand-alone shell script](b2488935d718ffc84965de047bdd474c2879db24).

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+          }
+        }
+        // 14. Opbeans
+        stage('Opbeans') {
```
This should be tag-based, this pipeline is master based IIRC, therefore the tag conditional can be left as it was previously, what do you think?

cachedout replied:
@v1v Do you mean this this conditional specifically?

```
      when {
        tag pattern: 'v\\d+\\.\\d+\\.\\d+', comparator: 'REGEXP'
      }
```

v1v replied:
Let me clarify, I meant, this particular stage could be kept in the other CI Jenkinsfile. IIUC, this particular pipeline will be triggered for the master branch and as a consequence, it will create a tag/branch with the version.

Therefore, the other CI Jenkinsfile could be kept as it was, it was configured to listen for branches/tags/PRs and the `when condition` in place will care of running that particular stage. 

I think it might help to simplify this pipeline. what do you think?

cachedout replied:
Gotcha! I agree and have restored the original logic. Thanks!

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+    DOCKERHUB_SECRET = 'secret/apm-team/ci/elastic-observability-dockerhub'
+    ELASTIC_DOCKER_SECRET = 'secret/apm-team/ci/docker-registry/prod'
+    NEXUS_SECRET = 'secret/apm-team/ci/nexus'
+    MAVEN_CONFIG = '-Dmaven.repo.local=.m2'
```
nit:
maybe DEFAULT_MAVEN_CONFIG ?

v1v replied:
I said nothing, that's how it's defined in the other Jenkinsfile!

cachedout replied:
Aye. I just moved it from there. :)

SylvainJuge commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+
+  }
+  parameters {
+    string(name: 'MAVEN_CONFIG', defaultValue: '-B -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn', description: 'Additional maven options.')
```
What are the use-cases of overriding maven config on a per build basis ? I'm pretty sure we haven't had to use this so far.

cachedout replied:
Very good question. I used [the original Jenkinsfile](https://github.com/elastic/apm-agent-java/blob/master/Jenkinsfile#L34) as a skeleton for this one and that's where this came from. I don't see where it would come into play in this workflow, however, so I think it is safe to remove.

cachedout replied:
I have consolidated this into a single step [in this commit](d17f51e1add8472b5c93290be2a1b75677161993).

SylvainJuge commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+              def INPUT_PARAMS = input message: 'Do you wish to update the version?', ok: 'Next', parameters: [choice(name: 'UPDATE_CHOICE', choices:['Yes', 'No'], description: "This will run `mvn release:update-versions`")]
+              if (INPUT_PARAMS == 'Yes') {
+                  dir("${BASE_DIR}") {
+                    sh mvn release:update-versions
```
we switched to using `versions` plugin instead in latest version of the release instructions https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#releasing

cachedout replied:
Ah! Glad you caught that. Thanks! I will make this change.

SylvainJuge replied:
As mentioned in my comment, the change was mostly due to having the `release` plugin to change formatting in `pom.xml`, and not just update versions.

cachedout replied:
I have updated this stage substantially. The new workflow is as follows:

1. Display the current version present in `pom.xml` and ask the user if they would like to do an update. If they select "no", then execution continues to the next stage.

2. If "yes" has been selected, an additional dialog appears again displaying the current version along with an input field for the user to enter a new version.

3. After submitting the new version, `mvn versions:set -DnewVersion=${new_version}` is run.

4. Execution continues from there.

SylvainJuge commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+        stage('Check oss.sonatype.org') {
+          steps {
+            // If this fails, an exception should be thrown and execution will halt
+            httpRequest(url: "https://oss.sonatype.org/#welcome")
```
does the `#welcome` part of the URL matters here ?

cachedout replied:
It doesn't matter. The root of the site redirects here so I just put it in to avoid the (slight) overhead of the redirect.

felixbarny replied:
For a future improvement: could we use this: https://status.maven.org/history.atom

cachedout replied:
A while ago I looked into using the Atom feed for this kind of thing but it's fraught with problems. Unfortunately, it's basically just free-text and there is no clear way to determine whether a given entry represents an ongoing outage, a recovery, an announcement, or free puppies. 🐶 I have some ideas about how to do this kind of thing in the future with Github since they have enabled webhooks. My current plan is to get that working and then see if Sonatype is open to the possibility of using webhooks on their page as well.

SylvainJuge commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+          steps {
+            deleteDir()
+            dir("${OPBEANS_REPO}"){
+              git credentialsId: 'f6c7695a-671e-4f4f-a331-acdce44ff9ba',
```
probably worth having a non-obscure credentials ID and/or a proper env variable for this.

cachedout replied:
Using this method, we can use Jenkins secure credentials storage to ensure that we don't leak credentials to the log. Could you please elaborate on your reason for wanting this not to be obscured?

SylvainJuge replied:
There is no special reason, I was just surprised to not have something referred by a human-friendly name here.

SylvainJuge commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+        stage('Docker push') {
+          when {
+            beforeAgent true
+            expression { return params.push_docker }
```
What controls `params.push_docker` value ? Isn't it always `true` for release pipeline ?

cachedout replied:
You are right. This came from the previous Jenkinsfile where it was gated, but that gate should not be necessary here. I will remove it. Thanks!

cachedout replied:
This has been removed. Thanks for catching this!

SylvainJuge commented:
There have been a few recent changes in [release instructions](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#releasing).
- updating changelog early in the process
- use `versions` maven plugin to update versions (this one does not re-formats `pom.xml` as `release` plugin does, which is annoying.
- waiting for documentation to be published, otherwise users will get a 404 when opening link from changelog

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+              def spid = getVault('nexus')["staging-profile-id"]
+              dir("${BASE_DIR}"){
+                withEnvMask(vars: [[var: "SPID", password: spid]]){
+                  def foundStagingId = nexusFindStagingId(stagingProfileId: "${SPID}", groupId: "co.elastic.apm")
```
what happens if there are multiple staging repos (for example a leftover from a failed release)?

cachedout replied:
At present, [we consider this an error requiring manual intervention](https://github.com/elastic/apm-pipeline-library/blob/master/vars/nexusFindStagingId.groovy#L68).

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+              dir("${BASE_DIR}"){
+                withEnvMask(vars: [[var: "SPID", password: spid]]){
+                  def foundStagingId = nexusFindStagingId(stagingProfileId: "${SPID}", groupId: "co.elastic.apm")
+                  nexusCloseStagingRepository(stagingProfileId: "${SPID}", stagingId: foundStagingId)
```
Closing can take a minute or two. Does this return after triggering the close or after the close is finished?

cachedout replied:
We attempt to poll the repo for the actual close event, so this should block until the close actually happens.

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+        // 10. Update cloudfoundry/index.yml
+        // Needs to append a line such as the following:
+        // 1.13.0: https://repo1.maven.org/maven2/co/elastic/apm/elastic-apm-agent/1.13.0/elastic-apm-agent-1.13.0.jar
+        stage('Update Cloudfoundry') {
```
let's to this after the `Wait for artifact to be available in Maven Central` step

cachedout replied:
I've moved this. Thanks!

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+                def dotX = sh(script: 'git tag|tail -1|sed s/v//|cut -f1 -d "."|awk \'{print $1".x"}\'', returnStdout: true)
+                def bareTag = sh(script: 'git tag|tail -1|sed s/v//', returnStdout: true)
+                def finalUrl = sh(script: "echo $baseUrl$dotX\.html#release-notes-$bareTag")
+                def ret = githubReleaseCreate(draft: true, body: "[Release Notes for $bareTag]($finalUrl)")
```
what's the title of the release? So far, we did `"Release $version"`.

cachedout replied:
Right now, we don't specify a name when we call the GitHub API but this is easy to change. I will make a change upstream in the APM Pipeline library to support this behavior.

cachedout replied:
[This change is now merged](https://github.com/elastic/apm-pipeline-library/pull/427). I will update this call to use the new functionality.

cachedout replied:
The call has been updated. Thanks for the suggestion!

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+                def baseUrl = "https://oss.sonatype.org/service/local/repositories/releases/content/co/elastic/apm/apm-agent-java"
+                def bareTag = sh(script: 'git tag|tail -1|sed s/v//', returnStdout: true)
+                def fullUrl = sh(script: "$baseUrl/$bareTag/apm-agent-java-$bareTag\.pom")
+                waitUntil(initialRecurrencePeriod: 15000) {
```
Does that mean it first starts to try after 15 sec? Is there a timeout? It usually takes around 10 min.

cachedout replied:
cURL has its own time-out, and after it expires, it will trigger the retry in the `waitUntil`. which will retry every 15 seconds, starting immediately, with a back-off. 

I think that we can improve this with the following changes:

1) Sleep first for ~10 minutes before we enter the re-try loop.
2) Bump the initial re-try up to 1 minute from 15s.

cachedout replied:
The above strategy is now implemented. Thanks for the feedback!

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+          }
+        }
+        // 3. Execute the release Jenkins job on the internal ci server.
+            // For now we assume that the release has been pushed to staging already on the internal CI
```
How does the release job get triggered?

cachedout replied:
By hand. I could probably improve this description but what I meant to indicate here is that this pipeline should be run after the existing release job is run.

felixbarny replied:
maybe open a dialog describing what to do which should be closed after the release job is done?

felixbarny replied:
BTW, I love that this just delegates to the existing and battle-tested release job instead of re-inventing this step which is definitely the most critical piece.

felixbarny replied:
Or we could trigger the release job when a new tag is created. That way, we don't have a direct dependency to actively trigger the build in the internal CI form the public CI.

The public CI  would execute the [`release:prepare`](http://maven.apache.org/maven-release/maven-release-plugin/examples/prepare-release.html) phase of the maven release plugin which updates the versions and creates a new tag.

The creation of the tag triggers the release job on the internal CI which executes the [`release:perform`](http://maven.apache.org/maven-release/maven-release-plugin/examples/perform-release.html) phase of the release plugin.

This does not have to, and probably shouldn't, be in the first iteration of the automated release.

cachedout replied:
I agree that this is a good long-term goal and I also agree that we shouldn't try to achieve this in the first iteration. Therefore, I'm going to resolve this comment because it's becoming a bit tricky to track all the feedback in this PR but we'll come back to it after we have this PR merged. Thanks, @felixbarny ! :)

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+          steps {
+            script {
+              // Show the select input modal
+              def INPUT_PARAMS = input message: 'Do you wish to update the version?', ok: 'Next', parameters: [choice(name: 'UPDATE_CHOICE', choices:['Yes', 'No'], description: "This will run `mvn release:update-versions`")]
```
Instead of updating the version here, I wonder if we should instead require the release version to be set in the release job. See https://maven.apache.org/maven-release/maven-release-plugin/examples/non-interactive-release.html

We could then also require to set the next dev version. The release plugin offers system properties for the release and next dev version: `-DreleaseVersion=1.2.0  -DdevelopmentVersion=1.3.0-SNAPSHOT`.

@eyalkoren @SylvainJuge WDYT?

SylvainJuge replied:
I'm not really sure to understand how this would be different: do we make the release job just "release current version", which is read from `pom.xml` ?

What if we keep having this prompt for release version and the default value is read from `pom.xml`, if a different value is set just updates version before executing release ?

Having this version set/bump done at CI level allows to avoid running any local git command, which is one less opportunity for human errors.

I'm also ok for doing an increment on the minor version number, but would rather use `mvn versions:set` and not `mvn:release` plugin. for the reason I mentioned above (which I can definitely get by).

felixbarny replied:
> but would rather use mvn versions:set and not mvn:release plugin

During a release:prepare and after a release:perform the versions get updated by the release plugin anyways.

So in the current state of this PR, we would update the version in step 2, then once again in on internal-ci via the `release:prepare` stage (it removes the `-SNAPSHOT`) and finally after the release is done, after the `release:perform` stage (increments the patch version and adds `-SNAPSHOT`, in most cases we actually want to increment the minor version).

What I'm suggesting is to modify the version 2 times instead of 3.

cachedout replied:
Hi @felixbarny and @SylvainJuge We've let this discussion sit for a while. I don't have any strong opinions on the way we handle this, but I'd like to come to a consensus so we can get this merged. What's the best way to find a path forward here?

felixbarny replied:
don't see this as blocking the PR, that's something we can always improve later on

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+            httpRequest(url: "https://oss.sonatype.org/#welcome")
+          }
+        }
+        // 2. Review project version.
```
Could we add a step that checks whether the current build on master is green?
(Not required for the first version)

cachedout replied:
Good idea! [I have introduced this capability upstream in this PR](https://github.com/elastic/apm-pipeline-library/pull/433).

cachedout replied:
I have introduced a step into this pipeline that warns if the master build is not passing and presents the user with an input asking if they would like to abort the build or continue.

eyalkoren commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+        // 12. Publish release on Github. This will notify users watching repository.
+        stage('Publish release on GitHub') {
+          steps {
+            githubReleasePublish(id: ${env.RELEASE_ID})
```
The GH release contains a link to a release-notes URL calculated in the `Create GitHub release draft` stage. This link will be available after the relevant branch gets built, after pushing the new tag in the `Branch creation` phase.
It takes time, so this step needs to block until the release-notes URL is available, otherwise we will publish a Release with a broken link.

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,235 @@
+        // 1. Check to see if oss.sonatype.org is up. We throw an exception and fail the job if there is no response.
+        stage('Check oss.sonatype.org') {
+          steps {
+            // If this fails, an exception should be thrown and execution will halt
```
We have added another step since you have started working on this regarding documentation:

> Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc) to reflect the new version release:
Go over PRs or git log and add bug fixes and features.
Move release notes from the Unreleased sub-heading to the correct [[release-notes-{major}.x]] sub-heading ([Example PR](https://github.com/elastic/apm-agent-java/pull/1027/files) for 1.13.0 release).

This step seems a bit tricky to automate. But we could instead just show this instruction in an input. When done with manually reviewing and adjusting the release notes, the automated release continues.

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,238 @@
+    NEXUS_SECRET = 'secret/apm-team/ci/nexus'
+    MAVEN_CONFIG = '-B -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn -Dmaven.repo.local=.m2'
+    OPBEANS_REPO = 'opbeans-java'
+
```
```suggestion
```

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,238 @@
+      environment {
+        HOME = "${env.WORKSPACE}"
+        JAVA_HOME = "${env.HUDSON_HOME}/.java/java10"
+        PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
```
These env variables are only defined in the top-level `Initializing` stage that it's only used in the nested checkout stage. Is this required for the other stages?

cachedout replied:
Good catch. I moved them up to the top level where they belong.

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,238 @@
+          steps {
+            dir("${BASE_DIR}") {
+              // This is required to ensure that Maven is working with a real branch
+              sh(script: "git checkout -f {$branch_specifier}")
```
```suggestion
              sh(script: "git checkout -f ${branch_specifier}")
```

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,238 @@
+            def ver = mvnVersion(showQualifiers: true)
+            // 2.2 Raise a prompt asking if the version is correct or if they wish to update
+            script {
+              def should_continue = input(message: "Current version is ${ver}", parameters: [
```
Moved to the script closure

```suggestion
            script {
              // 2.1 Fetch the current version from pom.xml
              def ver = mvnVersion(showQualifiers: true)
              // 2.2 Raise a prompt asking if the version is correct or if they wish to update
              def should_continue = input(message: "Current version is ${ver}", parameters: [
```

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,255 @@
+        stage('Checkout') {
+          steps {
+            dir("${BASE_DIR}") {
+              // This is required to ensure that Maven is working with a real branch
+              sh(script: "git checkout -f ${branch_specifier}")
```
A few comments regarding the git commands
- Is it possible to use the `gitCheckout` step instead the `sh` step? If so, then we can remove the `githubEnv` references 
- Can we use the `withGitRelease` step before pushing with the command `gitPush`?
- Can we use the `setupAPMGitEmail` step to configure the user.email and user.name?

cachedout replied:
I moved this into a shell script. Could you take another look?

apmmachine commented:
## :broken_heart: Build Failed 

<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1058/detail/PR-1058/58//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1058/detail/PR-1058/58//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1058/detail/PR-1058/58//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1058/detail/PR-1058/58//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-java_1058.docs-preview.app.elstc.co/diff) [![stats](https://img.shields.io/badge/stats-stats%20-blue)](https://ela.st/observabtl-ci-stats)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: [Pull request #1058 updated]
* Start Time: 2020-04-28T12:34:43.642+0000

* Duration: 12 min 26 sec (746218)


* Commit: b01df73088a7400d463e3f735647c35686e31c82



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 640  |
| Skipped      | 1 |
| Total        | 641   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: Shell Script
    * Description: #!/bin/bash                set -euxo pipefail                ./mvnw compile javadoc:javadoc                
    * Result: FAILURE
    
    * Duration: 2 min 34 sec<
    
    * Start Time: 2020-04-28T12:44:25.319+0000
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-04-28T12:46:28.265Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-04-28T12:46:28.274Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-04-28T12:46:29.611Z] [INFO] Codecov: Sending data...
[2020-04-28T12:46:29.687Z] [INFO] Codecov: Sending data...
[2020-04-28T12:46:29.701Z] [INFO] Codecov: Sending data...
[2020-04-28T12:46:30.062Z] + curl -s -o codecov.sh https://codecov.io/bash
[2020-04-28T12:46:30.175Z] + curl -s -o codecov.sh https://codecov.io/bash
[2020-04-28T12:46:30.175Z] + bash codecov.sh
[2020-04-28T12:46:30.175Z] 
[2020-04-28T12:46:30.175Z]   _____          _
[2020-04-28T12:46:30.175Z]  / ____|        | |
[2020-04-28T12:46:30.175Z] | |     ___   __| | ___  ___ _____   __
[2020-04-28T12:46:30.176Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2020-04-28T12:46:30.176Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2020-04-28T12:46:30.176Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2020-04-28T12:46:30.176Z]                               Bash-tbd
[2020-04-28T12:46:30.176Z] 
[2020-04-28T12:46:30.176Z] 
[2020-04-28T12:46:30.176Z] ==> Jenkins CI detected.
[2020-04-28T12:46:30.176Z]     project root: .
[2020-04-28T12:46:30.300Z] + curl -s -o codecov.sh https://codecov.io/bash
[2020-04-28T12:46:30.300Z] + bash codecov.sh
[2020-04-28T12:46:30.300Z] 
[2020-04-28T12:46:30.300Z]   _____          _
[2020-04-28T12:46:30.300Z]  / ____|        | |
[2020-04-28T12:46:30.300Z] | |     ___   __| | ___  ___ _____   __
[2020-04-28T12:46:30.300Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2020-04-28T12:46:30.300Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2020-04-28T12:46:30.301Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2020-04-28T12:46:30.301Z]                               Bash-tbd
[2020-04-28T12:46:30.301Z] 
[2020-04-28T12:46:30.301Z] 
[2020-04-28T12:46:30.301Z] ==> Jenkins CI detected.
[2020-04-28T12:46:30.518Z] --> token set from env
[2020-04-28T12:46:30.518Z]     Yaml not found, that's ok! Learn more at http://docs.codecov.io/docs/codecov-yaml
[2020-04-28T12:46:30.519Z] ==> Running gcov in . (disable via -X gcov)
[2020-04-28T12:46:30.519Z] ==> Python coveragepy not found
[2020-04-28T12:46:30.519Z] ==> Searching for coverage reports in:
[2020-04-28T12:46:30.519Z]     + .
[2020-04-28T12:46:30.563Z]     project root: .
[2020-04-28T12:46:30.563Z] --> token set from env
[2020-04-28T12:46:30.563Z]     Yaml not found, that's ok! Learn more at http://docs.codecov.io/docs/codecov-yaml
[2020-04-28T12:46:30.563Z] ==> Running gcov in . (disable via -X gcov)
[2020-04-28T12:46:30.563Z] ==> Python coveragepy not found
[2020-04-28T12:46:30.563Z] ==> Searching for coverage reports in:
[2020-04-28T12:46:30.563Z]     + .
[2020-04-28T12:46:32.261Z] --> No coverage report found.
[2020-04-28T12:46:32.261Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2020-04-28T12:46:32.264Z]     -> Found 20 reports
[2020-04-28T12:46:32.264Z] ==> Detecting git/mercurial file structure
[2020-04-28T12:46:32.264Z] ==> Reading reports
[2020-04-28T12:46:32.264Z]     + ./apm-agent-plugins/apm-jms-plugin/target/site/jacoco/jacoco.xml bytes=44585
[2020-04-28T12:46:32.264Z]     + ./apm-agent-plugins/apm-httpclient-core/target/site/jacoco/jacoco.xml bytes=5083
[2020-04-28T12:46:32.264Z]     + ./apm-agent-plugins/apm-mongoclient-plugin/target/site/jacoco/jacoco.xml bytes=12681
[2020-04-28T12:46:32.265Z]     + ./apm-agent-plugins/apm-es-restclient-plugin/apm-es-restclient-plugin-6_4/target/site/jacoco/jacoco.xml bytes=9248
[2020-04-28T12:46:32.265Z]     + ./apm-agent-plugins/apm-es-restclient-plugin/apm-es-restclient-plugin-common/target/site/jacoco/jacoco.xml bytes=12744
[2020-04-28T12:46:32.265Z]     + ./apm-agent-plugins/apm-jdbc-plugin/target/site/jacoco/jacoco.xml bytes=78883
[2020-04-28T12:46:32.265Z]     + ./apm-agent-plugins/apm-hibernate-search-plugin/apm-hibernate-search-plugin-6_x/target/site/jacoco/jacoco.xml bytes=5918
[2020-04-28T12:46:32.265Z]     + ./apm-agent-plugins/apm-hibernate-search-plugin/apm-hibernate-search-plugin-5_x/target/site/jacoco/jacoco.xml bytes=6020
[2020-04-28T12:46:32.265Z]     + ./apm-agent-plugins/apm-jaxws-plugin/target/site/jacoco/jacoco.xml bytes=4575
[2020-04-28T12:46:32.265Z]     + ./apm-agent-plugins/apm-jaxrs-plugin/target/site/jacoco/jacoco.xml bytes=18753
[2020-04-28T12:46:32.266Z]     + ./apm-agent-plugins/apm-asynchttpclient-plugin/target/site/jacoco/jacoco.xml bytes=14063
[2020-04-28T12:46:32.266Z]     + ./apm-agent-plugins/apm-redis-plugin/apm-redis-common/target/site/jacoco/jacoco.xml bytes=3102
[2020-04-28T12:46:32.266Z]     + ./apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/target/site/jacoco/jacoco.xml bytes=24314
[2020-04-28T12:46:32.266Z]     + ./apm-agent-plugins/apm-redis-plugin/apm-jedis-plugin/target/site/jacoco/jacoco.xml bytes=7365
[2020-04-28T12:46:32.266Z]     + ./apm-agent-plugins/apm-jmx-plugin/target/site/jacoco/jacoco.xml bytes=40668
[2020-04-28T12:46:32.266Z]     + ./apm-agent-plugins/apm-apache-httpclient-plugin/target/site/jacoco/jacoco.xml bytes=41207
[2020-04-28T12:46:32.266Z]     + ./apm-agent-plugins/apm-api-plugin/target/site/jacoco/jacoco.xml bytes=83750
[2020-04-28T12:46:32.267Z]     + ./apm-agent-plugins/apm-error-logging-plugin/target/site/jacoco/jacoco.xml bytes=10002
[2020-04-28T12:46:32.267Z]     + ./apm-agent-plugins/apm-java-concurrent-plugin/target/site/jacoco/jacoco.xml bytes=9173
[2020-04-28T12:46:32.267Z]     + ./apm-agent-core/target/site/jacoco/jacoco.xml bytes=941056
[2020-04-28T12:46:32.267Z] ==> Appending adjustments
[2020-04-28T12:46:32.267Z]     http://docs.codecov.io/docs/fixing-reports
[2020-04-28T12:46:32.267Z]     -> No adjustments found
[2020-04-28T12:46:32.268Z] ==> Gzipping contents
[2020-04-28T12:46:32.268Z] ==> Uploading reports
[2020-04-28T12:46:32.268Z]     url: https://codecov.io
[2020-04-28T12:46:32.268Z]     query: branch=cachedout%2Frelease_pipeline&commit=d2fcd00f30f265fd4c0ce7acce26db3d0ff4643b&build=58&build_url=https%3A%2F%2Fapm-ci.elastic.co%2Fjob%2Fapm-agent-java%2Fjob%2Fapm-agent-java-mbp%2Fjob%2FPR-1058%2F58%2F&name=&tag=&slug=elastic%2Fapm-agent-java&service=jenkins&flags=&pr=1058&job=
[2020-04-28T12:46:32.268Z]     -> Pinging Codecov
[2020-04-28T12:46:32.268Z] https://codecov.io/upload/v4?package=bash-tbd&token=secret&branch=cachedout%2Frelease_pipeline&commit=d2fcd00f30f265fd4c0ce7acce26db3d0ff4643b&build=58&build_url=https%3A%2F%2Fapm-ci.elastic.co%2Fjob%2Fapm-agent-java%2Fjob%2Fapm-agent-java-mbp%2Fjob%2FPR-1058%2F58%2F&name=&tag=&slug=elastic%2Fapm-agent-java&service=jenkins&flags=&pr=1058&job=
[2020-04-28T12:46:32.268Z]     -> Uploading
[2020-04-28T12:46:33.534Z]     -> View reports at https://codecov.io/github/elastic/apm-agent-java/commit/d2fcd00f30f265fd4c0ce7acce26db3d0ff4643b
[2020-04-28T12:46:36.302Z] Failed in branch Unit Tests
[2020-04-28T12:46:36.340Z] Failed in branch Smoke Tests 02
[2020-04-28T12:47:03.516Z] + bash codecov.sh
[2020-04-28T12:47:03.516Z] bash: codecov.sh: No such file or directory
[2020-04-28T12:47:03.516Z] + echo 'codecov exited with 127'
[2020-04-28T12:47:03.516Z] codecov exited with 127
[2020-04-28T12:47:07.329Z] Failed in branch Smoke Tests 01
[2020-04-28T12:47:07.577Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2020-04-28T12:47:07.635Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-04-28T12:47:07.660Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-04-28T12:47:09.143Z] Running on Jenkins in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1058
[2020-04-28T12:47:09.197Z] [INFO] getVaultSecret: Getting secrets
[2020-04-28T12:47:09.272Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-04-28T12:47:10.125Z] + chmod 755 generate-build-data.sh
[2020-04-28T12:47:10.125Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1058/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1058/runs/58 FAILURE 746218
[2020-04-28T12:47:10.676Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1058/runs/58/steps/?limit=10000 -o steps-info.json
[2020-04-28T12:47:11.469Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1058/runs/58/tests/?status=FAILED -o tests-errors.json
[2020-04-28T12:47:11.880Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1058/runs/58/log/ -o pipeline-log.txt

```

</p>
</details>



<!--PIPELINE
{"commit":"b01df73088a7400d463e3f735647c35686e31c82","number":"58","status":"FAILURE","url":"https://apm-ci.elastic.co/job/apm-agent-java/job/apm-agent-java-mbp/job/PR-1058/58/"}
PIPELINE-->

cachedout commented:
Hi friends. :)

This should now be ready for a final round of reviews. Hopefully I've addressed all the concerns but if there is any unresolved issue please let me know and we can either address it here or file an issue for a follow-up. Thanks!

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,229 @@
+        stage('Branch creation') {
+          steps {
+            dir("${BASE_DIR}") {
+              setupAPMGitEmail()
```
See my previous comment

```suggestion
```

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,229 @@
+                credentialsId: 'f6c7695a-671e-4f4f-a331-acdce44ff9ba',
+                shallow: false
+            )
+            stash(allowEmpty: true, name: 'source', useDefaultExcludes: false)
+            dir("${BASE_DIR}") {
+              setupAPMGitEmail()
+              script {
+                env.TAG_BARE = sh(script: "git tag | tail -1", returnStdout: true)
+                env.TAG_VER = sh(script: "git tag | tail -1 | sed s/v//", returnStdout: true)
+                env.TAG_DOT_X = sh(script: 'git tag|tail -1|sed s/v//|cut -f1 -d "."|awk \'{print $1".x"}\'', returnStdout: true)
+              }
+            }
```
What do you think to move the setupAPMGitEmail before the stash? Then all the other calls afteeward are not required:

```suggestion
            dir("${BASE_DIR}") {
              setupAPMGitEmail()
            }
            stash(allowEmpty: true, name: 'source', useDefaultExcludes: false)
            dir("${BASE_DIR}") {
              script {
                env.TAG_BARE = sh(script: "git tag | tail -1", returnStdout: true)
                env.TAG_VER = sh(script: "git tag | tail -1 | sed s/v//", returnStdout: true)
                env.TAG_DOT_X = sh(script: 'git tag|tail -1|sed s/v//|cut -f1 -d "."|awk \'{print $1".x"}\'', returnStdout: true)
              }
            }
```

cachedout replied:
I was wondering about that. I was just being extra-cautious here since there isn't much harm in setting it multiple times but if it will work this way than that's even better.

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,229 @@
+            deleteDir()
+            unstash 'source'
+            dir("${BASE_DIR}"){
+              script {
+                env.BRANCH_NAME = "master"
+              }
+              withGitRelease() {
+                setupAPMGitEmail()
```
See my previous comments
```suggestion
              withGitRelease() {
```

v1v commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,229 @@
+          steps {
+            dir("${BASE_DIR}") {
+              setupAPMGitEmail()
+              script {
+                env.BRANCH_NAME = "tags/${env.TAG_BARE}"
+              }
+              withGitRelease() {
+                script {
+                  def curVer = mvnVersion()
+                  sh(script: "./scripts/jenkins/branch_creation.sh ${curVer}")
+                  def isMajor = sh(script: "./scripts/jenkins/is_major.sh ${curVer}", returnStatus: true)
+                  if (isMajor == 0) {
+                    input message: "This was a major version release. Please update the conf.yml in the docs repo before continuing", ok "Continue"
+                  }
+                }
+                gitPush()
+              } 
```
IIRC, `BRANCH_NAME` is a read only env variable, `withEnv` could help to change the value for the release context, what do you think?

```suggestion
              withEnv(["BRANCH_NAME=tags/${env.TAG_BARE}"]){
                withGitRelease() {
                  script {
                    def curVer = mvnVersion()
                    sh(script: "./scripts/jenkins/branch_creation.sh ${curVer}")
                    def isMajor = sh(script: "./scripts/jenkins/is_major.sh ${curVer}", returnStatus: true)
                    if (isMajor == 0) {
                      input message: "This was a major version release. Please update the conf.yml in the docs repo before continuing", ok "Continue"
                    }
                  }
                  gitPush()
                }
              }
              
```

cachedout replied:
Interesting. I had no problems in setting it directly when I was testing it locally but I agree that if we can use a context-based approach that it is much cleaner. Thanks for pointing this out. I wasn't familiar with `withEnv`.

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,229 @@
+                    ]
+                  ]
+                )
+                sh(name: "mavenVersionUpdate", script: "mvn versions:set -DnewVersion=${new_version}")
```
An issue with `versions:set` is that it creates a `pom.xml.versionsBackup` file for every `pom.xml` in the project.

I think `mvn --batch-mode release:update-versions -DdevelopmentVersion=${new_version}` would be the better option.

Also, I don't see the new changes being committed in this step, is that correct?

cachedout replied:
Thanks, @felixbarny. I have updated the `mvn` call per your suggestion. I have also added a `git commit`. Let me know if it looks good to you.

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,226 @@
+              }
+            }
+          }
+        }
```
Add a step that just prints a message like.

`Start the release job on the internal CI. Click proceed once the job has succeeded. Click cancel if the release has failed and manually undo the release.`

cachedout replied:
Good call! Step added.

felixbarny commented:
The build failure in the Javadoc step should be fixed once you merge the master in your branch.

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,234 @@
+                    [
+                      $class: 'StringParameterDefinition',
+                      defaultValue: "${ver}",
+                      description: 'We will run mvn versions:set -DnewVersion=<NEW_VERSION>', name: 'New Version'
```
```suggestion
                      description: 'We will update the project version in all pom.xml files. Set this to your desired <release-version>-SNAPSHOT (for example 1.2.3-SNAPSHOT if you want to release version 1.2.3).', name: 'New Version'
```

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,234 @@
+              def should_continue = input(message: "Current version is ${ver}", parameters: [
+                [
+                  $class: 'ChoiceParameterDefinition',
+                  name: "Do you wish to update the version?",
```
```suggestion
                  name: "You are about to release version ${ver - '-SNAPSHOT'}. Do you wish to update the version?",
```

felixbarny commented on .ci/release/Jenkinsfile:
```diff
@@ -0,0 +1,234 @@
+                )
+                sh(name: "mavenVersionUpdate", script: "mvn --batch-mode release:update-versions -DdevelopmentVersion=${new_version}")
+                withGitRelease() {
+                  sh(script: "git commit -a -m 'Version bump'")
```
don't forget to push :)

cachedout replied:
The reason I didn't push here is that it happens in the step shortly after. Do you think we should have a push here as well? I'm fine with either. :)

felixbarny replied:
If we don't push here, the internal CI doesn't see the changes

cachedout replied:
Ah! You're right of course. I'll make this change.

cachedout commented:
Just an update on this so people don't think it's abandoned. From my perspective, this is just awaiting a final review from @eyalkoren and @SylvainJuge . No hurry whatsoever, of course, but I wanted to make sure everybody is on the same page so there's no confusion. Thanks!

cachedout commented:
Tests are passing and so I am merging this [per the most recent comment by Felix](https://github.com/elastic/apm-agent-java/pull/1058#issuecomment-641994056). 

We can schedule a pair-release session whenever you like. Thanks all for the feedback and direction on this!
